File: store/108.将有序数组转换为二叉搜索树.c
/*
 * @lc app=leetcode.cn id=108 lang=c
 *
 * [108] 将有序数组转换为二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

struct TreeNode *atoBST(int *nums, int low, int high)
{
  if (low > high)
    return NULL;
  // int mid = (low + high) / 2;
  int mid = low + (high - low) / 2;
  struct TreeNode *node = (struct TreeNode *)malloc(sizeof(struct TreeNode));
  node->val = nums[mid];
  node->left = atoBST(nums, low, mid - 1);
  node->right = atoBST(nums, mid + 1, high);
  return node;
}

struct TreeNode *sortedArrayToBST(int *nums, int numsSize)
{
  // 法一：递归
  // return atoBST(nums, 0, numsSize - 1);
  // 法二：迭代
  struct TreeNode **stack, *node, *root, *temp;
  int top1, top2, top3, low, high, mid, *left, *right;
  stack = (struct TreeNode **)malloc(sizeof(struct TreeNode *) * 1000);
  left = (int *)malloc(sizeof(int) * 1000);
  right = (int *)malloc(sizeof(int) * 1000);
  top1 = top2 = top3 = 0;
  root = (struct TreeNode *)malloc(sizeof(struct TreeNode));
  root->left = NULL;
  root->right = NULL;
  stack[top1++] = root;
  left[top2++] = 0;
  right[top3++] = numsSize - 1;
  while (top1 > 0)
  {
    node = stack[--top1];
    low = left[--top2];
    high = right[--top3];
    mid = (low + high) / 2;
    node->val = nums[mid];
    if (low <= mid - 1)
    {
      temp = (struct TreeNode *)malloc(sizeof(struct TreeNode));
      temp->val = nums[mid];
      temp->left = NULL;
      temp->right = NULL;
      node->left = temp;
      stack[top1++] = node->left;
      left[top2++] = low;
      right[top3++] = mid - 1;
    }
    if (high >= mid + 1)
    {
      temp = (struct TreeNode *)malloc(sizeof(struct TreeNode));
      temp->val = nums[mid];
      temp->left = NULL;
      temp->right = NULL;
      node->right = temp;
      stack[top1++] = node->right;
      left[top2++] = mid + 1;
      right[top3++] = high;
    }
  }
  return root;
}
// @lc code=end

File: store/108.将有序数组转换为二叉搜索树.js
/*
 * @lc app=leetcode.cn id=108 lang=javascript
 *
 * [108] 将有序数组转换为二叉搜索树
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} nums
 * @return {TreeNode}
 */
const getBST = (nums, left, right) => {
  if (left > right) return null;
  let mid = Math.floor((left + right) / 2);
  // if (mid < 0 || mid > nums.length - 1) return;
  let node = new TreeNode(nums[mid]);
  node.left = getBST(nums, left, mid - 1);
  node.right = getBST(nums, mid + 1, right);
  return node;
};
var sortedArrayToBST = function (nums) {
  //递归
  // return getBST(nums, 0, nums.length - 1);
  // 迭代
  let queue = [],
    leftQueue = [],
    rightQueue = [],
    front,
    rear,
    mid;
  front = rear = 0;
  let root = new TreeNode();
  queue[rear++] = root;
  leftQueue[rear - 1] = 0;
  rightQueue[rear - 1] = nums.length - 1;
  while (front !== rear) {
    let curNode = queue[front++];
    let leftIndex = leftQueue[front - 1];
    let rightIndex = rightQueue[front - 1];
    let mid = Math.floor((leftIndex + rightIndex) / 2);
    curNode.val = nums[mid];

    if (leftIndex < mid) {
      queue[rear++] = new TreeNode();
      curNode.left = queue[rear - 1];
      leftQueue[rear - 1] = leftIndex;
      rightQueue[rear - 1] = mid - 1;
    }

    if (rightIndex > mid) {
      queue[rear++] = new TreeNode();
      curNode.right = queue[rear - 1];
      leftQueue[rear - 1] = mid + 1;
      rightQueue[rear - 1] = rightIndex;
    }
  }

  return root;
};
// @lc code=end
